services/src/ogc/wms/request.rs

```diff
@@ -0,0 +1,199 @@
+    // ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub enum GetFeatureInfoFormat {
+    TextXML //, ...
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetStyles {
+    pub version: String,
+    pub layer: String,
+}
+
+#[derive(PartialEq, Debug, Deserialize, Serialize)]
+pub struct GetLegendGraphic {
+    pub version: String,
+    pub layer: String,
+    // ...
+}
+
+pub fn parse_bbox<'de, D>(deserializer: D) -> Result<BoundingBox2D, D::Error>
+    where D: serde::Deserializer<'de>
+{
+    // format is : "x1,y1,x2,y2"
+
+    let s = <&str as serde::Deserialize>::deserialize(deserializer)?;
+
+    let split: Vec<Result<f64, std::num::ParseFloatError>> = s.split(',')
+        .map(str::parse).collect();
+
+    if let [Ok(x1), Ok(y1), Ok(x2), Ok(y2)] = *split.as_slice() {
+        BoundingBox2D::new(
+            Coordinate2D::new(x1, y1),
+            Coordinate2D::new(x2, y2))
+            .map_err(|_| D::Error::custom("Not a valid bbox"))
+    } else {
+        Err(D::Error::custom("Not a valid bbox"))
+    }
+}
+
+pub fn parse_time<'de, D>(deserializer: D) -> Result<Option<TimeInterval>, D::Error>
+    where D: serde::Deserializer<'de>
+{
+    // time is specified in ISO8601, it can either be an instant (single datetime) or an interval
+    // An interval is separated by "/". "Either the start value or the end value can be omitted to
+    // indicate no restriction on time in that direction."
+    // sources: - http://docs.geoserver.org/2.8.x/en/user/services/wms/time.html#wms-time
+    //          - http://www.ogcnetwork.net/node/178
+
+    // TODO: support relative time intervals
+
+    let s = <&str as serde::Deserialize>::deserialize(deserializer)?;
+
+    let split: Vec<_> = s.split('/').map(|s| chrono::DateTime::parse_from_rfc3339(s)).collect();
+
+    match *split.as_slice() {
+        [Ok(time)] => TimeInterval::new(time.timestamp(), time.timestamp())
+            .map(Some)
+            .map_err(|_| D::Error::custom("Not a valid time")),
+        [Ok(start), Ok(end)]  => TimeInterval::new(start.timestamp(), end.timestamp())
+            .map(Some)
+            .map_err(|_| D::Error::custom("Not a valid time")),
+        _ => Err(D::Error::custom("Not a valid time"))
+    }
+}
+
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use geoengine_datatypes::primitives::{BoundingBox2D, Coordinate2D};
+
+    #[test]
+    fn deserialize_get_map() {
+        let query = "request=GetMap&service=WMS&version=1.3.0&layer=test&bbox=1,2,3,4&width=2&height=2&crs=foo&styles=ssss&format=image/png&time=2000-01-01T00:00:00.0Z/2000-01-02T00:00:00.0Z";
+        let parsed: WMSRequest = serde_urlencoded::from_str(query).unwrap();
+
+        let request = WMSRequest::GetMap(GetMap {
+            version: "1.3.0".into(),
+            width: 2,
+            layer: "test".into(),
+            crs: "foo".to_string(),
+            styles: "ssss".to_string(),
```

Comment:
Decide for one transformation at least within a single function.

It is the same in the other test.

Reply:
139e2fba1ff1f1d9e93408358e758a05c15f2f12